Smart Connector UIExtension

yentelmanero opened the pull request:
This PR is part of my bachelor research conducted in TU Wien in the [Business Informatics Group](https://www.big.tuwien.ac.at/).

---

In this PR, the Smart Connector UIExtension, which is a Tool that hovers around the selected node, simplifying the creation of new nodes and edges, was created.

Here are some screenshots of the UIExtension:

![image](https://github.com/eclipse-glsp/glsp-client/assets/71039676/2f4b3857-6532-41ec-b71e-5f014b62d378)
![image](https://github.com/eclipse-glsp/glsp-client/assets/71039676/a4d6ad59-6c98-4578-826c-ee1e62e63812)
(different types of nodes can have different options)

[Server PR](https://github.com/eclipse-glsp/glsp-server-node/pull/70)
[Theia PR](https://github.com/eclipse-glsp/glsp-theia-integration/pull/188)

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector-module.ts:
```diff
@@ -0,0 +1,33 @@
+/********************************************************************************
+ * Copyright (c) 2023 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
```
```suggestion
/********************************************************************************
 * Copyright (c) 2023 Business Informatics Group (TU Wien) and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
```

haydar-metin commented on packages/client/src/utils/html-utils.ts:
```diff
@@ -23,3 +23,29 @@ export function createElementFromHTML(html: string): HTMLElement | undefined {
     return undefined;
 }
+
+declare global {
+    interface HTMLElement {
+        next(): HTMLElement
+        previous(): HTMLElement
+        first(): HTMLElement
+        last(): HTMLElement
+    }
+}
+
+// HTMLElement extensions for readability and convenience (reduce casting)
+HTMLElement.prototype.next = function (): HTMLElement {
+    return this.nextElementSibling as HTMLElement;
+};
+
+HTMLElement.prototype.previous = function (): HTMLElement  {
+    return this.previousElementSibling as HTMLElement;
+};
+
+HTMLElement.prototype.first = function (): HTMLElement {
+    return this.firstElementChild as HTMLElement;
+};
+
+HTMLElement.prototype.last = function (): HTMLElement  {
+    return this.lastElementChild as HTMLElement;
+};
```
It seems reasonable to me, but let us wait for what others will say.

yentelmanero replied:
I used this a lot, so reusing this cast was tedious.

haydar-metin commented on packages/protocol/src/action-protocol/smart-connector.ts:
```diff
@@ -0,0 +1,65 @@
+/********************************************************************************
+ * Copyright (c) 2021-2023 STMicroelectronics and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
```
```suggestion
/********************************************************************************
 * Copyright (c) 2023 Business Informatics Group (TU Wien) and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
```

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+/********************************************************************************
+ * Copyright (c) 2021-2023 STMicroelectronics and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
```
```suggestion
/********************************************************************************
 * Copyright (c) 2023 Business Informatics Group (TU Wien) and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
```

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+            this.createTool(item);
+        }
+        if (event.ctrlKey) {
+        // matchesKeystroke with Ctrl and Ctrl+F does not seem to work on Windows 11/Chrome
```
Reminder to find out why afterward

haydar-metin reviewed (commented):
Thanks for the PR! I did just a preliminary look through the code. After the server-side compiles correctly, I will try it out and look deeper into the code.

Can you also add some description to your PR and reference the other PRs you have created, something like:

```
This PR is part of my bachelor research conducted in TU Wien in the [Business Informatics Group](https://www.big.tuwien.ac.at/).

---

<Little Description>

<Other PR references>
```

yentelmanero commented:
> Thanks for the PR! I did just a preliminary look through the code. After the server-side compiles correctly, I will try it out and look deeper into the code.
> 
> Can you also add some description to your PR and reference the other PRs you have created, something like:
> 
> ```
> This PR is part of my bachelor research conducted in TU Wien in the [Business Informatics Group](https://www.big.tuwien.ac.at/).
> 
> ---
> 
> <Little Description>
> 
> <Other PR references>
> ```

@haydar-metin, unfortunately, the server cannot build because of the missing packages from this PR. Is there a way to circumvent this issue?

haydar-metin commented on packages/protocol/src/action-protocol/smart-connector.ts:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * The identifier of the element where the smart connector is to be opened.
+     */
+    selectedElementID: string;
```
```suggestion
    selectedElementId: string;
```

The code base uses `Id` usually.

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+import { IDiagramStartup } from '../../base/model/diagram-loader';
+import { EnableDefaultToolsAction } from '../../base/tool-manager/tool';
+
+const CONTAINER_PADDING_PX = 20;
+const MAX_HEIGHT_GROUP = '150px';
+const SEARCH_FIELD_SUFFIX = '_search_field';
+const SMART_CONNECTOR_CONTAINER_ID = 'smart-connector-container';
+const EXPAND_BUTTON_ID = 'smart-connector-expand-button';
+const GROUP_CONTAINER_CLASS = 'smart-connector-group-container';
+const HEADER_CLASS = 'smart-connector-group-header';
+const GROUP_CLASS = 'smart-connector-group';
+const TOOL_BUTTON_CLASS = 'smart-connector-button';
+const COLLAPSABLE_CLASS = 'collapsable-group';
```
```suggestion
export namespace SmartConnector {
export const CONTAINER_PADDING_PX = 20;
export const MAX_HEIGHT_GROUP = '150px';
export const SEARCH_FIELD_SUFFIX = '_search_field';
export const SMART_CONNECTOR_CONTAINER_ID = 'smart-connector-container';
export const EXPAND_BUTTON_ID = 'smart-connector-expand-button';
export const GROUP_CONTAINER_CLASS = 'smart-connector-group-container';
export const HEADER_CLASS = 'smart-connector-group-header';
export const GROUP_CLASS = 'smart-connector-group';
export const TOOL_BUTTON_CLASS = 'smart-connector-button';
export const COLLAPSABLE_CLASS = 'collapsable-group';
}
```

If we export it as suggested here, then others can also reuse the values.

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        this.hideSmartConnector();
+    }
+
+    protected async requestAvailableOptions(contextElement?: GModelElement): Promise<void> {
```
This method does more than one thing:
1) initializes the Options
2) creates the body

for this reason, i would say, move `createBody`to `onBeforeShow` and rename this method to `initAvailableOptions`

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        containerElement.setAttribute('aria-label', 'Smart-Connector');
+    }
+
+    protected createBody(): void {
```
```suggestion
    protected initBody(): void {
```
you are not creating a new element. You are initializing the body. Create would mean, that you are returning something.

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        this.smartConnectorContainer = smartConnectorContainer;
+    }
+
+    protected createExpandButton(): void {
```
```suggestion
    protected initExpandButton(): void {
```

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+
+    protected createSearchField(itemGroup: SmartConnectorGroupItem): HTMLElement {
+        const searchField = document.createElement('input');
+        searchField.classList.add('smart-connector-search');
```
Please extract that string to the `consts` defined before as you already did in other places here.

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        searchField.onkeydown = ev => this.searchFieldKeyHandler(ev, itemGroup);
+        this.searchFields[itemGroup.id] = searchField;
+        const searchContainer = document.createElement('div');
+        const containerClass = itemGroup.submenu ? 'smart-connector-submenu-search-container' : 'smart-connector-search-container';
```
Please extract those strings to the `consts` defined before as you already did in other places here.

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+
+    // #region event handlers
+
+    protected requestFilterUpdate(filter: string, itemGroup: SmartConnectorGroupItem): void {
```
Im not sure if hiding or having only visible elements in the DOM is better. Lets discuss that later.

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        }
+    }
+
+    protected toolButtonKeyHandler(event: KeyboardEvent, item: PaletteItem): void {
```
```suggestion
    protected handleToolButtonKey(event: KeyboardEvent, item: PaletteItem): void {
```

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        this.toolButtonNavigation(event, item);
+    }
+
+    protected searchFieldKeyHandler(event: KeyboardEvent, itemGroup: SmartConnectorGroupItem): void {
```
```suggestion
    protected handleSearchFieldKey(event: KeyboardEvent, itemGroup: SmartConnectorGroupItem): void {
```

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+
+    // #region navigation handlers
+
+    protected searchFieldNavigation(event: KeyboardEvent, itemGroup: SmartConnectorGroupItem): void {
```
```suggestion
    protected navigateSearchField(event: KeyboardEvent, itemGroup: SmartConnectorGroupItem): void {
```

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        }
+    }
+
+    protected headerNavigation(event: KeyboardEvent, groupItems: HTMLElement, header: HTMLElement): void {
```
```suggestion
    protected navigateHeader(event: KeyboardEvent, groupItems: HTMLElement, header: HTMLElement): void {
```

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        }
+    }
+
+    protected toolButtonNavigation(event: KeyboardEvent, item: PaletteItem): void {
```
```suggestion
    protected navgiateToolButton(event: KeyboardEvent, item: PaletteItem): void {
```

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        };
+    }
+
+    protected createTool(item: PaletteItem): void {
```
```suggestion
    protected triggerCreation(item: PaletteItem): void {
```

Or something similar?

haydar-metin commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -0,0 +1,524 @@
+        const smartConnector = expandButton?.parentElement;
+        if (matchesKeystroke(event, 'Space') && smartConnector?.style.visibility === 'visible'
+            && expandButton?.style.visibility === 'visible') {
+            showSmartConnector(container, expandButton);
```
I'm not sure if it is okay how the extension is expanded. You are currently hiding it with CSS and then make it visible again with CSS.

I think the whole state change (circle -> expanded) should be handled in the class directly. I would say, create a new `Action` with the state you want e.g., collapsed, expanded as a parameter and trigger it here. Then let your `UIExtension` handle the action and depending on the provided `State` update it to be either `collapsed` or `expanded`.

So if you click space, you would tell your `UIExtension` that it should move from the circle view to the box view.

yentelmanero replied:
You're right, I am not sure why I initially did it this way.

I added the new Action now, but I did not add a key binding to close the Smart Connector with (since I am not sure which keys I can use so they don't overlap), just Space to open it.

haydar-metin reviewed (commented):
Thank you very much, I just tested it and reviewed the code. Largely the code looks good. There are some small requests from me. Functionality wise it looks very good! Just small remarks.

Can you also add the finished solution as an image to the PR? Similar to the images provided here.

## Styling
![image](https://github.com/eclipse-glsp/glsp-client/assets/13104167/3c1abaec-8fc3-4fc1-8b35-bf7cb61fcddd)

I think visually it would be more appealing if the boxes would have a top and bottom border. Otherwise it looks like there is more.

## Usage
![image](https://github.com/eclipse-glsp/glsp-client/assets/13104167/1bf55d75-db93-47d7-bc4e-542b8e635e79)

I think in those cases nothing should be shown?

---

We are almost there! Good job!

haydar-metin commented:
## UI

![image](https://github.com/eclipse-glsp/glsp-client/assets/13104167/3732bf8f-0e11-4c7a-80ee-befac40dda10)

1. You probably need to add a background. `.smart-connector-group-container` is transparent
2. The arrow is not correctly aligned
3. Scrollbar is not visible - in the CSS you want to show it but is not shown

Can you add a listener for `Escape` that closes the smart connector

## Other

1. Please resolve all the comments I have written and which are already finished (I can not to that)
2. Please install [Prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) and run the auto format for all of your files
3. I will then notify someone else and then we can probably merge it if everything works as intended

yentelmanero commented:
> ## UI
> ![image](https://private-user-images.githubusercontent.com/13104167/303315464-3732bf8f-0e11-4c7a-80ee-befac40dda10.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc0MTAzNDMsIm5iZiI6MTcwNzQxMDA0MywicGF0aCI6Ii8xMzEwNDE2Ny8zMDMzMTU0NjQtMzczMmJmOGYtMGUxMS00YzdhLTgwZWUtYmVmYWM0MGRkYTEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA4VDE2MzQwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUxYTRhYTZiYjhkM2M3ZmM5YzExMDhmMDcxNWJlZGI1YThmYmQxM2M3MDMzMjg4YzM5MzBhOGJjODA5N2ZlYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Iku6ha8IOHdVuQyJSmyQuChnpwU7WcS_3Ax5RdGCzKY)
> 
> 1. You probably need to add a background. `.smart-connector-group-container` is transparent
> 2. The arrow is not correctly aligned
> 3. Scrollbar is not visible - in the CSS you want to show it but is not shown
> 
> Can you add a listener for `Escape` that closes the smart connector
> 
> ## Other
> 1. Please resolve all the comments I have written and which are already finished (I can not to that)
> 2. Please install [Prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) and run the auto format for all of your files
> 3. I will then notify someone else and then we can probably merge it if everything works as intended

Hi, how are you running the project? I am running the client and then the theia integration and there are no issues here as you can see in a similar image:

![image](https://github.com/eclipse-glsp/glsp-client/assets/71039676/e3e63371-9254-48e9-823c-7ef1ab4f8a0f)

haydar-metin commented:
Hey, I also tested the `standalone` version, that means you need to build the `client` with `yarn` and start your node server with `yarn start -w -p 8081`. Then you need to open the file: `file:///.../glsp-client/examples/workflow-standalone/app/diagram.html` in your browser. I think you just need to move some of your CSS there.

https://github.com/eclipse-glsp/glsp-client?tab=readme-ov-file#how-to-start-the-workflow-diagram-example

yentelmanero commented on packages/client/src/features/smart-connector/smart-connector.ts:
```diff
@@ -366,11 +381,25 @@ export class SmartConnector extends AbstractUIExtension implements IActionHandle
+            this.hideSmartConnector();
+            // assumes that the graph is always the last child of base div
+            // this focus is done to "reactivate" the key listener to re-open if needed
+            document.getElementById(this.options.baseDiv)?.last().focus();
```
Since the focus is lost when the connector is closed, the KeyListener does not react to any inputs. It does also not react when any tool is selected. Because of this, I just added a onkeydown event on every tool and when it closes, it refocuses on the base div (the graph). Is this assumption here correct? Or is there a better way to get the base div?

yentelmanero replied:
Because the KeyListener does not react anymore. So if I, for example, open the connector with space, close it again with escape and then try to open it again, it won't work

yentelmanero commented:
@haydar-metin Hi, is there any information regarding the status of the PR? Are there any more changes to be done?

haydar-metin commented:
@yentelmanero I requested a second review internally. After that, we can merge it.

tortmayr commented on packages/client/src/default-modules.ts:
```diff
@@ -99,7 +100,8 @@ export const DEFAULT_MODULES = [
     svgMetadataModule,
-    statusModule
+    statusModule,
+    smartConnectorModule
```
Please also open a PR against https://github.com/eclipse-glsp/glsphttps://github.com/eclipse-glsp/glsp
and add the `SelectionPalette` to the feature list (as part of the contexts section,, marked as experimental)

tortmayr reviewed (changes requested):
In general the code looks good to me and the `SmartConnector` seems to work really well.
I have a couple of inline comments/suggestions.

As we discussed offline we would prefer to drop the concept of `SmartConnector` in favor of a more generic `SelectionPalette` and `SelectionPaletteProvider` API.

Please rename the corresponding artifacts accordingly.

Also there seems to be some room for improvements regarding the positing of the smartconnector when moving/resizing a node and scrolling/zooming:

- Scrolling/zooming: When scrolling or zooming the smart connector gets hidden although the node is still selected
- Resize: The position of the `+` button is not recentered when resizing
- Moving: The smart connector gets hidden during move (which is fine imo) but is not set visible again once the move is finished. (Node is still selected)

However, these are issue that can be addressed in a follow-up PR and don't necessarily have to be covered by you.

yentelmanero commented:
Hi @tortmayr, thanks for the feedback.

I changed the code snippets requested and also added that the selection palette does not disappear when panning and zooming, since that changed seemed simple enough to do it relatively quickly.

Since I changed quite a bit of code there, because I am not using the DOM measurements anymore but the Viewport and GNode data, I would ask you to please review those parts once again. The reason I changed that is because element.getBoundingClientRect() was too slow in delivering the needed information, so I would calculate with the old frame, while already showing a new frame, resulting in wrong behaviour when zooming.

I am also not entirely sure if the way I added the explicit module imports (for both workflow-standalone and workflow-theia) was done correctly, but it worked after I added those and it did not before.

yentelmanero commented:
> Hi @tortmayr, thanks for the feedback.
> 
> I changed the code snippets as requested and also added that the selection palette does not disappear when panning and zooming, since that changed seemed simple enough to do it relatively quickly.
> 
> Since I changed quite a bit of code there, because I am not using the DOM measurements anymore but the Viewport and GNode data, I would ask you to please review those parts once again. The reason I changed that is because element.getBoundingClientRect() was too slow in delivering the needed information, so I would calculate with the old frame, while already showing a new frame, resulting in wrong behaviour when zooming.
> 
> I am also not entirely sure if the way I added the explicit module imports (for both workflow-standalone and workflow-theia) was done correctly, but it worked after I added those and it did not before.

Update: now also added the selection palette moving when resizing or moving nodes, since I could not find an explanation why I did not do it when writing my thesis. Please review

tortmayr commented:
@yentelmanero 
Thank your for this awesome contribution.
I have reviewed and tested the changes. Everything seems to work as expected and all review comments have been addressed.
Nice work!

We will keep this PR open and merge it after the 2.2.0 Release is done (end of May). This way we have a full release cycle to polish the feature and address follow-up issues before we are shipping it to adopters.

In any case, from our side this concludes the practical aspect of your bachelor thesis. We don't require/desire any more changes from your side. Congrats!
